patterns/interpreter.md

```diff
@@ -0,0 +1,317 @@
+term -> term / factor
+term -> factor
+factor -> ( exp )
+factor -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
+```
+
+In fact this grammar should be further transformed
+depending on what we are going to do with the grammar.
+For example, we might need to remove left recursion.
+For more details please see
+[Compilers:Principles,Techniques, and Tools
+](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools)
+(aka Dragon Book).
+
+## Solution 1
+
+Our first approach is a standard one, simple implementation of
+a recursive descent parser. The following code
+doesn't have `struct` abstraction in order to keep code short.
+The code panics when expression is syntactically wrong
+(unbalanced parentheses or missing digit/operator for example).
+
+```rust
+fn token(input: &[u8], cur: usize) -> char {
+    if cur < input.len() {
+        input[cur] as char
+    } else {
+        '$' // End of line
+    }
+}
+
+fn single_expr(input: &[u8], cur: &mut usize, out: &mut Vec<String>) {
+    expr(input, cur, out);
+
+    let ch = token(input, *cur);
+    if ch != '$' {
+        panic!("Unexpected symbol '{}', at {}", ch, *cur);
+    }
+}
+
+fn expr(input: &[u8], cur: &mut usize, out: &mut Vec<String>) {
+    term(input, cur, out);
+
+    loop {
+        let ch = token(input, *cur);
+        if ch == '$' || (ch != '+' && ch != '-') {
+            break;
+        } else {
+            *cur += 1;
+            term(input, cur, out);
+            translate(ch, out);
+        }
+    }
+}
+
+fn term(input: &[u8], cur: &mut usize, out: &mut Vec<String>) {
+    factor(input, cur, out);
+
+    loop {
+        let ch = token(input, *cur);
+        if ch == '$' || (ch != '*' && ch != '/') {
+            break;
+        } else {
+            *cur += 1;
+            factor(input, cur, out);
+            translate(ch, out);
+        }
+    }
+}
+
+fn factor(input: &[u8], cur: &mut usize, out: &mut Vec<String>) {
+    let ch = token(input, *cur);
+
+    if ch.is_digit(10) {
+        out.push(format!("push {}", ch));
+    } else if ch == '(' {
+        *cur += 1;
+        expr(input, cur, out);
+
+        let ch = token(input, *cur);
+        if ch != ')' {
+            panic!("Missing ')' at {}", *cur);
+        }
+    } else {
+        panic!("Unexpected symbol '{}', at {}", ch, *cur);
+    }
+
+    *cur += 1;
+}
+
+fn translate(ch: char, out: &mut Vec<String>) {
+    out.push(String::from("pop ebx"));
+    out.push(String::from("pop eax"));
+    out.push(format!("{} eax, ebx", to_oper(ch)));
+    out.push(String::from("push eax"));
+}
+
+fn to_oper(ch: char) -> String {
+    match ch {
+        '+' => return String::from("add"),
+        '-' => return String::from("sub"),
+        '*' => return String::from("mul"),
+        '/' => return String::from("div"),
+        _ => panic!("Invalid operator '{}'", ch),
+    }
+}
+
+pub fn main() {
+    let mut out = vec![];
+    let mut cur = 0;
+    let exp = b"2/(7-3)";
+
+    single_expr(exp, &mut cur, &mut out);
```

Comment:
you are not using cur after this, so there is no need to take it as argument.
It could be created inside `single_expr`

Reply:
@MarcoIeni  In fact I think the best solution is to wrap everything into a struct `Parser`, define `index` and other necessary vars like `index`, and hence we could git rid of passing around `index`. I didn't t want to bloat the code.